core/tx_list.go

```diff
@@ -696,18 +696,27 @@ func (l *txPricedList) Underpriced(tx *types.Transaction) bool {
 }
 
 func (l *txPricedList) underpricedForMulti(h *multiCurrencyPriceHeap, tx *types.Transaction) bool {
-	underpriced := l.underpricedFor(h.nativeCurrencyHeap, tx)
+	// Has to be underpriced for ALL heaps to be underpriced for the whole pool.
+	if !l.underpricedFor(h.nativeCurrencyHeap, tx, h.IsCheaper) {
```

Comment:
I suggest we don't use `underpricedFor` at all, since we have to cancel out the check for the empty heap anyway AND also pass the external `IsCheaper` comparison function. It makes sense to do the opposite and extract the "pruning" part into a separate function and just call it.